File: Algorithm_I_study_plan/house_robber.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        elif len(nums) == 2:
            return max(nums[0], nums[1])

        nums[1] = max(nums[0], nums[1])

        for i in range(2, len(nums)):
            nums[i] = max(nums[i-1], nums[i] + nums[i-2])

        return nums[-1]

File: NeetCode_Roadmap/house_robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size() == 1) return nums[0];

        int skip_next = 0;
        int next = nums.back();
        for (int i = nums.size() - 2; i >= 0; --i)
        {
            int tmp = next;
            next = max(nums[i] + skip_next, next);
            skip_next = tmp;
        }
        return max(next, skip_next);
    }
};